### ISSUE TITLE
False positive: undefined class xpath selector

### ISSUE THREAD

**Issue by funivan** [vimeo/psalm#2006]
Do not know why Psalm check string and report about undefined class
```php
ERROR: UndefinedClass
src/Test.php:35:43 
Class or interface //div[@class="product"]/parent does not exist
foreach ($paginator->getDetailLinks('//div[@class="product"]/parent::a/@href') as $url) {
```
I'm using `xpath` to fetch data from the page. And every time I write selector with `parent::a` I get and error from the psalm.

**Comment by weirdan** [comment]
I was able to reproduce this only when the `getDetailLink` is annotated as `@param class-string $p`: https://psalm.dev/r/ab147ee219

@funivan if that's the case, you probably want to use ordinary `string`, not `class-string` for the `getDetailLink()` parameter.

**Comment by funivan** [comment]
I can reproduce this issue:
```php
<?php
/** @param string|callable $p */
function f($p): array {
  return [];
}
foreach (f('//div[@class="product"]/parent::a/@href') as $url) {
}
```
https://psalm.dev/r/d5372e9dd6

**Comment by muglug** [comment]
Simplified to:

```php
/** @param string|callable $p */
function f($p): array {
  return [];
}
foreach (f('#b::a') as $url) {}
```
https://psalm.dev/r/a1b7add736

**Comment by muglug** [comment]
Fixed in 669b50e
